crates/blockchain-tree/src/blockchain_tree.rs

```diff
@@ -309,16 +309,57 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTree<DB, C, EF>
 
         // check if block parent can be found in Tree
         if let Some(chain_id) = self.block_indices.get_blocks_chain_id(&parent.hash) {
+            // get the canonical fork of the chain so we can unwind later
+            let canonical_fork = match self.canonical_fork(chain_id) {
+                None => {
+                    return Err(InsertBlockError::tree_error(
+                        BlockchainTreeError::BlockSideChainIdConsistency { chain_id },
+                        block.block,
+                    ))
+                }
+                Some(fork) => fork,
+            };
+
+            // now we unwind
+            let old_canon_chain = self
+                .revert_canonical(canonical_fork.number)
+                .map_err(|err| InsertBlockError::new(block.block.clone(), err.into()))?;
+
             // found parent in side tree, try to insert there
-            return self.try_insert_block_into_side_chain(block, chain_id)
+            let res = self.try_insert_block_into_side_chain(block.clone(), chain_id);
+
+            // re insert canonical if possible
+            if let Some(old_chain) = old_canon_chain {
+                self.commit_canonical(old_chain)
+                    .map_err(|err| InsertBlockError::new(block.block.clone(), err.into()))?;
+            }
+
+            trace!(target: "blockchain_tree", ?res, "Inserted block into side chain");
+
+            return res
         }
 
         // if not found, check if the parent can be found inside canonical chain.
         if self
             .is_block_hash_canonical(&parent.hash)
             .map_err(|err| InsertBlockError::new(block.block.clone(), err.into()))?
         {
-            return self.try_append_canonical_chain(block)
+            // now we unwind
+            let old_canon_chain = self
+                .revert_canonical(parent.number)
+                .map_err(|err| InsertBlockError::new(block.block.clone(), err.into()))?;
+
+            // found parent in side tree, try to insert there
```

Comment:
`side tree` -> `canonical chain`?

And if we append the tip of the canonical, don't need unwind and recommit?